dotnet/csharp.bzl

```diff
@@ -48,50 +48,56 @@ def _get_libdirs(dlls, libdirs=[]):
   return [dep.dirname for dep in dlls] + libdirs
 
 def _make_csc_arglist(ctx, output, depinfo, extra_refs=[]):
+  args = ctx.actions.args()
+
   flag_start = ctx.attr._flag_start
-  args = [
-       # /out:<file>
-      _make_csc_flag(flag_start, "out", output.path),
-       # /target (exe for binary, library for lib, module for module)
-      _make_csc_flag(flag_start, "target", ctx.attr._target_type),
-      # /fullpaths
-      _make_csc_flag(flag_start, "fullpaths"),
-      # /warn
-      _make_csc_flag(flag_start, "warn", str(ctx.attr.warn)),
-      # /nologo
-      _make_csc_flag(flag_start, "nologo"),
-  ]
+
+  # /out:<file>
+  args.add(_make_csc_flag(flag_start, "out", output.path))
+
+  # /target (exe for binary, library for lib, module for module)
+  args.add(_make_csc_flag(flag_start, "target", ctx.attr._target_type))
+
+  # /fullpaths
+  args.add(_make_csc_flag(flag_start, "fullpaths"))
+
+  # /warn
+  args.add(_make_csc_flag(flag_start, "warn", str(ctx.attr.warn)))
+
+  # /nologo
+  args.add(_make_csc_flag(flag_start, "nologo"))
 
   # /modulename:<string> only used for modules
   libdirs = _get_libdirs(depinfo.dlls)
   libdirs = _get_libdirs(depinfo.transitive_dlls, libdirs)
 
   # /lib:dir1,[dir1]
   if libdirs:
-    args += [_make_csc_flag(flag_start, "lib", ",".join(list(libdirs)))]
+    args.add(_make_csc_flag(flag_start, "lib", ",".join(list(libdirs))))
```

Comment:
You can join these lazily and save analysis-time memory:

```
def _comma_join(flag, values):
  return _make_csc_flag(None, flag, ",".join(list(libdirs)))

(...)

  if libdirs:
    args.add((flag_start + "lib", libdirs), map_fn=_comma_join)
```

Reply:
Awesome! Will do x2

Reply:
I went with separate args instead because I had trouble with doing this (join all with commas and prefix that with `/foo:`) with a scalar `map_fn` (I think that's what you were going for in your example - where the value is a tuple.)

`before_each` won't work because it inserts a separate argument between each argument, but csc/mcs wants this to be a prefix of each argument (e.g. no space/newline)

`format` is simple to read but does result in larger param files. In practice the number of `lib` and `reference` args shouldn't be too crazy. In our build we have ~1500 `csharp_library` targets and with this change our total param file size went from around 13.1MB (lol) to 13.9MB.